File: 0 - 1 Knapsack Problem.py
class Solution:
    def helper(self, W, wt, val, dp, idx):
        if idx == len(wt): return 0
        if wt[idx] > W: return self.helper(W, wt, val, dp, idx + 1)
        if dp[idx][W] != -1: return dp[idx][W]
        
        dp[idx][W] = max(
            val[idx] + self.helper(W - wt[idx], wt, val, dp, idx + 1),
            self.helper(W, wt, val, dp, idx + 1)
        )
        
        return dp[idx][W]
        
    def knapSack(self,W, wt, val):
        return self.helper(W, wt, val, [[-1 for _ in range(W + 1)] for _ in range(len(wt))], 0)

File: 0 - 1 Knapsack Problem.java
class Solution {
    static int knapsack(int W, int val[], int wt[]) {
        int dp[]=new int[W+1];
        for(int i=0;i<val.length;i++)
            for(int w=W;w>=wt[i];w--)
                dp[w]=Math.max(dp[w],val[i]+dp[w-wt[i]]);
        return dp[W];
        
    }
}
